### PR TITLE
NXP backend: Condition Linear+BatchNorm fusing passes to valid combination only

### PR DESCRIPTION
**Pull request by StrycekSimon**
### Summary

Introduce shape conditioning to Linear+BN fusion related passes. 
Fix `RemoveSimulatedLinearBatchNormFusionQATPass` pass recompiling graph when not necessary.

**Problem description:**
Linear and BatchNorm layers are fusable only in cases where both layers work with the same dimension. Linear layer works strictly with only the last dimension of it's input. BatchNorm on the other hand works in channels dimension (the second one in most cases) except for one case when using BatchNorm1d and input shape `[N, C]` (see [official PyTorch documentation](https://docs.pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html) for more info). In this one case both BatchNorm and Linear works on the same dimension and are correctly fusable.

**Example:**
```python
inp = torch.randn((2, 2, 3)) # (N, C, L)
lin = nn.Linear(in_features=3, out_features=5) # Weight of shape (5, 3)
bn = nn.BatchNorm2d(num_features=2) # Running var shape (2, )

# With the following fusing equation:
# linear_w_fused = linear_w * (gamma / sqrt(var + eps))
# And the following shapes:
# linear_w.shape: (5, 3); var.shape: (2, )
# We cannot correctly perform the multiplication.
```

### Test plan
New unit test cases were added + relevant older ones were adjusted.

cc @robert-kalmar @JakeStevens @digantdesai

### CODE REVIEW

File: backends/nxp/tests/models.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 import torch
 
+from sqlalchemy import false
+
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/tests/models.py]: Don't know how did that get there. I removed it.

File: backends/nxp/tests/models.py [hunk 2]
```diff
@@ -457,51 +459,110 @@ def forward(self, x):
...
-            self.bn = torch.nn.BatchNorm1d(64, affine=bn_affine)
-        elif conv_module == "conv2d_t":
-            self.conv = torch.nn.ConvTranspose2d(3, 64, 3, padding=1, bias=conv_bias)
-            self.bn = torch.nn.BatchNorm2d(64, affine=bn_affine)
+class ConvBatchNormModule(torch.nn.Module):
+    def __init__(
+        self,
+        bias: bool,
+        input_rank: int,
+        num_features: int,
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/tests/models.py:490]: Good point! I fixed that.

**Comment by StrycekSimon** [comment]
The lintrunner fail refers to `test/end2end/test_end2end.py` file, which is unrelated to this PR. The failed tests seems all unrelated too.

**Comment by MartinPavella** [comment]
In the PR description you provide an example with:
```python
inp = torch.randn((2, 2, 2)) # (N, C, L)
lin = nn.Linear(in_features=3, out_features=5) # Weight of shape (5, 3)
```
How can you have 3 input channels for the `Linear` when the input does not have 3 channels?

File: backends/nxp/backend/graph_utils.py [hunk 3]
```diff
@@ -36,3 +36,9 @@ def is_op_node(node: Node, target_op) -> bool:
 
 def is_batch_norm(node: Node) -> bool:
     return is_op_node(node, batch_norm_target_ops)
+
+
+def get_output_shape(node: Node) -> torch.Size | None:
+    if node.meta.get("tensor_meta") is not None:
+        return node.meta["tensor_meta"].shape
```
Comment (MartinPavella) [inline backends/nxp/backend/graph_utils.py]: Why use `tensor_meta` instead of `val`?
I can see `tensor_meta` used only a handful of times in the entire code base. `val` is used much more often. What is the difference?

File: backends/nxp/tests/ir/edge_passes/test_linear_bn_fusing.py [hunk 4]
```diff
@@ -258,3 +266,49 @@ def test_linear_bn_full_qat_pipeline_conversion(
...
+        ops=[
+            exir_ops.edge.aten.addmm.default,
+            exir_ops.edge.aten.linear.default,
+        ],
+    )
+    assert graph_contains_any_of_ops(
+        graph=edge_program.graph,
+        ops=batch_norm_target_ops,
+    )
+    assert not any("lowered_module" in node.name for node in edge_program.graph.nodes)
```
Comment (MartinPavella) [inline backends/nxp/tests/ir/edge_passes/test_linear_bn_fusing.py]: We prefer to compare for `torch.ops.higher_order.executorch_call_delegate` target now, to avoid string comparisons:
```python
assert not graph_contains_any_of_ops(edge_program.graph, [torch.ops.higher_order.executorch_call_delegate])
```

**Comment by StrycekSimon** [comment]
> In the PR description you provide an example with:
> 
> ```python
> inp = torch.randn((2, 2, 2)) # (N, C, L)
> lin = nn.Linear(in_features=3, out_features=5) # Weight of shape (5, 3)
> ```
> 
> How can you have 3 input channels for the `Linear` when the input does not have 3 channels?

That was a typoo.

**Review by MartinPavella — APPROVED** [review]

File: backends/nxp/backend/graph_utils.py (hunk 3 above)
Comment (StrycekSimon) [reply backends/nxp/backend/graph_utils.py]: Both should be correct. I will change it to `val` for consistency.

File: backends/nxp/tests/ir/edge_passes/test_linear_bn_fusing.py (hunk 4 above)
Comment (StrycekSimon) [reply backends/nxp/tests/ir/edge_passes/test_linear_bn_fusing.py]: Fixed. ✅

File: backends/nxp/backend/graph_utils.py [hunk 5]
```diff
@@ -36,3 +36,9 @@ def is_op_node(node: Node, target_op) -> bool:
 
 def is_batch_norm(node: Node) -> bool:
     return is_op_node(node, batch_norm_target_ops)
+
+
+def get_output_shape(node: Node) -> torch.Size | None:
+    if node.meta.get("val") is not None:
+        return node.meta["val"].shape
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/backend/graph_utils.py]: I added support for both.

**Comment by StrycekSimon** [comment]
I fixed the issue that Copilot mentioned. Can you please merge after the CI passes @MartinPavella?

**Comment by MartinPavella** [comment]
> I fixed the issue that Copilot mentioned. Can you please merge after the CI passes @MartinPavella?

I would (the changes look good), but the PR has conflicts. Please resolve them first.

File: backends/nxp/tests/models.py [hunk 6]
```diff
@@ -463,51 +462,114 @@ def forward(self, x):
...
-        elif conv_module == "conv2d_t":
-            self.conv = torch.nn.ConvTranspose2d(3, 64, 3, padding=1, bias=conv_bias)
-            self.bn = torch.nn.BatchNorm2d(64, affine=bn_affine)
+        if (input_rank == 2 or input_rank == 3) and not transposed_conv:
+            self.conv = torch.nn.Conv1d(
+                in_channels=num_features,
+                out_channels=num_features,
+                kernel_size=3,
+                bias=bias,
+            )
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/tests/models.py:503]: Incorrect. Torch 1D convs do support (N, L) inputs and it is clearly stated in the documentation.

File: backends/nxp/backend/graph_utils.py [hunk 7]
```diff
@@ -36,3 +36,14 @@ def is_op_node(node: Node, target_op) -> bool:
...
     return is_op_node(node, batch_norm_target_ops)
+
+
+def get_output_shape(node: Node) -> tuple[torch.Size] | torch.Size | None:
+    val = node.meta.get("val")
+
+    if isinstance(val, torch.Tensor):
+        return val.shape
+    elif isinstance(val, tuple):
+        return tuple([v.shape for v in val])
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/backend/graph_utils.py:47]: Let's keep it and merge.

**Review by roman-janik-nxp — APPROVED** [review]

**Comment by StrycekSimon** [comment]
I resolved the failing test by skipping it during PTQ. Fun fact: In PTQ, TorchAO [does Linear+BN fusing](https://github.com/pytorch/ao/blob/1d75a0779f2e0095a08b3753f4a4f1b9666ccfe0/torchao/quantization/pt2e/quantize_pt2e.py#L106) by itself but misses the conditioning introduced in this PR. This is most probably a bug in TorchAO.

Update: I created an [issue](https://github.com/pytorch/ao/issues/4116#issuecomment-4091867641) in TorchAO repository.

**Review by robert-kalmar — APPROVED** [review]
